src/components/shared/thread/SelectedActivityGraph.js

```diff
@@ -0,0 +1,351 @@
+    return callNodesOnSameCategoryPath[callNodesOnSameCategoryPath.length - 1];
+  }
+  return clickedCallNode;
+}
+
+function getSampleCategories(
+  samples: SamplesTable,
+  stackTable: StackTable
+): Array<IndexIntoSamplesTable | null> {
+  return samples.stack.map(s => (s !== null ? stackTable.category[s] : null));
+}
+
+class SelectedThreadActivityGraphCanvas extends PureComponent<Props> {
+  _onSampleClick = (sampleIndex: IndexIntoSamplesTable) => {
+    const {
+      fullThread,
+      filteredThread,
+      callNodeInfo,
+      selectedThreadIndex,
+      changeSelectedCallNode,
+      focusCallTree,
+    } = this.props;
+    const unfilteredStack = fullThread.samples.stack[sampleIndex];
+    if (unfilteredStack === null) {
+      return;
+    }
+
+    const clickedCategory = fullThread.stackTable.category[unfilteredStack];
+    const { callNodeTable, stackIndexToCallNodeIndex } = callNodeInfo;
+    const sampleCallNodes = getSampleCallNodes(
+      filteredThread.samples,
+      stackIndexToCallNodeIndex
+    );
+    const clickedCallNode = sampleCallNodes[sampleIndex];
+    if (clickedCallNode === null) {
+      return;
+    }
+
+    const sampleCategories = getSampleCategories(
+      fullThread.samples,
+      fullThread.stackTable
+    );
+    const chosenCallNode = findBestCallNode(
+      callNodeInfo,
+      sampleCallNodes,
+      sampleCategories,
+      clickedCallNode,
+      clickedCategory
+    );
+    // Change selection twice: First, to clickedCallNode, in order to expand
+    // the whole call path. Then, to chosenCallNode, to get the large-area
+    // graph highlighting.
+    changeSelectedCallNode(
+      selectedThreadIndex,
+      getCallNodePathFromIndex(clickedCallNode, callNodeTable)
+    );
+    changeSelectedCallNode(
+      selectedThreadIndex,
+      getCallNodePathFromIndex(chosenCallNode, callNodeTable)
+    );
+    focusCallTree();
+  };
+  render() {
+    const {
+      fullThread,
+      filteredThread,
+      interval,
+      rangeStart,
+      rangeEnd,
+      callNodeInfo,
+      selectedCallNodeIndex,
+      categories,
+      samplesSelectedStates,
+      treeOrderSampleComparator,
+    } = this.props;
+
+    return (
+      <div>
+        <ThreadActivityGraph
+          interval={interval}
+          fullThread={fullThread}
+          className="selectedThreadActivityGraph"
+          rangeStart={rangeStart}
+          rangeEnd={rangeEnd}
+          onSampleClick={this._onSampleClick}
+          categories={categories}
+          samplesSelectedStates={samplesSelectedStates}
+          treeOrderSampleComparator={treeOrderSampleComparator}
+        />
+        <ThreadStackGraph
+          interval={interval}
+          thread={filteredThread}
+          className="selectedThreadStackGraph"
+          rangeStart={rangeStart}
+          rangeEnd={rangeEnd}
+          callNodeInfo={callNodeInfo}
+          selectedCallNodeIndex={selectedCallNodeIndex}
+          onSampleClick={this._onSampleClick}
```

Comment:
Maybe we should have two different `onSampleClick` handlers for the two graphs. The `findBestCallNode` behavior is appropriate for the activity graph (it highlights roughly what you'd expect, once you're used to it), but it's less appropriate for clicking the stack graph, I've found. In the stack graph, I still expect clicking a sample to select the leaf / self node at that sample.